Draft: Download manager

cyberphantom52 opened the pull request:
Implements bottlesdevs/next-download-manager#1

TODO:

- [x] Retry with backoff
- [x] Cancel Download
- [ ] Pause/Resume
- [ ] Callbacks (Do we even need these)
- [ ] Timeouts
- [ ] Better progress tracking
- [ ] DownloadManager stats(e.g active/queued/failed/completed downloads,  total data downloaded) 
- [ ] ...

cyberphantom52 committed:
chore: Add reqwest crate

cyberphantom52 committed:
downloader: Implement DownloadManager

mirkobrombin commented on src/downloader.rs:
I see the status channel being created but never updated to InProgress, Completed, or Failed (am I wrong?)
There are also more things that are missing but I imagine this is still a WIP.

Good job for the moment <3

cyberphantom52 replied:
This might not even be working :). I just wanted to put it out so the design can evolve with comments.

mirkobrombin commented on src/downloader.rs:
```diff
@@ -0,0 +1,154 @@
+            progress: progress_tx,
+        };
+
+        let _ = self.queue.try_send(req);
```
Is the try_send safe to be ignored here?

cyberphantom52 replied:
nope, i just need a new error type to represent the error here.

mirkobrombin replied:
Roger!

cyberphantom52 committed:
error: Add reqwest errors

cyberphantom52 committed:
downloader: Propagate errors and correctly update download status

cyberphantom52 committed:
downloader: Implement cancelling downloads

cyberphantom52 committed:
downloader: Implement retries

gageberz commented on src/downloader.rs:
```diff
@@ -0,0 +1,211 @@
+    }
+}
+
+async fn download_thread(client: Client, req: &DownloadRequest) -> Result<File, Error> {
+    let mut resp = client.get(req.url.as_ref()).send().await?;
+    let total_bytes = resp.content_length();
+    let mut bytes_downloaded = 0u64;
+    let mut cancelled = false;
+    let mut file = File::options()
+        .read(true)
+        .write(true)
+        .create(true)
+        .open(&req.destination)?;
+
+    while let Some(chunk) = resp.chunk().await.transpose() {
+        cancelled = !req.cancel.is_empty();
+        if cancelled {
+            break;
+        }
+        let chunk = chunk?;
+        file.write_all(&chunk)?;
+        bytes_downloaded += chunk.len() as u64;
+        let _ = req.status.send(Status::InProgress(DownloadProgress {
+            bytes_downloaded,
+            total_bytes,
+        }));
+    }
+
+    if cancelled {
+        std::fs::remove_file(&req.destination)?;
+        Err(Error::Io(std::io::Error::new(
+            std::io::ErrorKind::Interrupted,
+            "Download cancelled",
+        )))
+    } else {
+        // Reset the cursor to the beginning of the file
+        file.seek(SeekFrom::Start(0))?;
+        Ok(file)
+    }
+}
```
```suggestion
async fn download_thread(client: Client, req: &mut DownloadRequest) -> Result<File, Error> {
    let mut resp = client
        .get(req.url.clone())
        .send()
        .await?
        .error_for_status()?;
    let total_bytes = resp.content_length();
    let mut bytes_downloaded = 0u64;
    let mut file = File::options()
        .read(true)
        .write(true)
        .create(true)
        .truncate(true)
        .open(&req.destination)?;

    req.status
        .send(Status::InProgress(DownloadProgress {
            bytes_downloaded,
            total_bytes,
        }))
        .ok();

    loop {
        tokio::select! {
            _ = &mut req.cancel => {
                eprintln!("Download of {} cancelled.", req.url);
                std::fs::remove_file(&req.destination)?;
                return Err(Error::Io(std::io::Error::new(
                            std::io::ErrorKind::Interrupted,
                            "Download cancelled",
                        )));
            }
            chunk_result = resp.chunk() => {
                match chunk_result {
                    Ok(Some(chunk)) => {
                        file.write_all(&chunk)?;
                        bytes_downloaded += chunk.len() as u64;
                        let _ = req.status.send(Status::InProgress(DownloadProgress {
                            bytes_downloaded,
                            total_bytes,
                        }));
                    }
                    Ok(None) => {
                        file.seek(SeekFrom::Start(0))?;
                        return Ok(file);
                    }
                    Err(e) => {
                        return Err(Error::Reqwest(e));
                    }
                }
            }
        }
    }
}

```

cyberphantom52 committed:
downloader: Improve File handling and use switch to tokio::fs::File

This is to make sure we don't block the runtime.

cyberphantom52 committed:
downloader: Use tokio::select to check for cancellations

cyberphantom52 committed:
downloader: Send initial progress update and return  error for response

cyberphantom52 committed:
downloader: Implement Future trait for DownloadHandle

This is the idomatic way

cyberphantom52 commented on src/error.rs:
```diff
@@ -8,4 +8,6 @@ pub enum Error {
     Serde(#[from] serde_json::Error),
     #[error("Reqwest: {0}")]
     Reqwest(#[from] reqwest::Error),
+    #[error("Oneshot: {0}")]
+    Oneshot(#[from] tokio::sync::oneshot::error::RecvError),
```
I don't like having to add this specific error type.

mirkobrombin replied:
Maybe, not sure, `RecvError` could map to a more general channel cancellation variant instead of adding a dedicated Oneshot error

cyberphantom52 replied:
I guess we can add `anyhow` crate and don't have to worry about this

cyberphantom52 commented on src/downloader.rs:
```diff
@@ -0,0 +1,236 @@
+                        })).ok();
+                    }
+                    Ok(None) => break,
+                    Err(e) => return Err(Error::Reqwest(e)),
```
need to remove_file in this case too

cyberphantom52 committed:
downloader: End progress updates at completion too

cyberphantom52 committed:
downloader: Manually drop file handle before deleting on error or cancel

cyberphantom52 commented on src/downloader.rs:
```diff
@@ -0,0 +1,236 @@
+        manager
+    }
+
+    pub fn set_max_parallel_downloads(&self, limit: usize) {
+        let current = self.semaphore.available_permits();
+        if limit > current {
+            self.semaphore.add_permits(limit - current);
+        } else if limit < current {
+            let to_remove = current - limit;
+            for _ in 0..to_remove {
+                let _ = self.semaphore.try_acquire();
+            }
+        }
+    }
```
On looking carefully, This has a race condition so this needs to be reimplemented.

edfloreshz committed:
fix: gptk impl

- added missing impl for GPTK
- fixed warnings and clippy

cyberphantom52 committed:
downloader: Refactor retry logic and improve errors

cyberphantom52 committed:
downloader: Add method to wait for status updates

cyberphantom52 committed:
downloader: fix retry logic

cyberphantom52 committed:
downloader: Use tokio_util::CancellationToken to handle cancellations

This fixes a bug where if we didn't store the handle to the download somewhere it would automatically cancel the download

cyberphantom52 committed:
downloader: Add hierarchical cancel tokens to cancel all downloads

cyberphantom52 commented on src/downloader.rs:
```diff
@@ -0,0 +1,286 @@
+    }
+}
+
+async fn dispatcher_thread(
```
dispatch and download threads can outlive the `DownloadManager` itself. Is this desirable or should we keep track of the threads to manage their lifecycles?

cyberphantom52 committed:
downloader: Split the implementation into a module

cyberphantom52 commented on src/downloader/worker.rs:
```diff
@@ -0,0 +1,138 @@
+    let mut file = File::create(&req.destination).await?;
+
+    update_progress(bytes_downloaded, total_bytes);
+    loop {
+        tokio::select! {
+            _ = req.cancel.cancelled() => {
+                drop(file); // Manually drop the file handle to ensure that deletion doesn't fail
+                tokio::fs::remove_file(&req.destination).await?;
+                return Err(Error::Download(DownloadError::Cancelled));
+            }
+            chunk = response.chunk() => {
+                match chunk {
+                    Ok(Some(chunk)) => {
+                        file.write_all(&chunk).await?;
+                        bytes_downloaded += chunk.len() as u64;
+                        update_progress(bytes_downloaded, total_bytes);
+                    }
+                    Ok(None) => break,
+                    Err(e) => {
+                        drop(file); // Manually drop the file handle to ensure that deletion doesn't fail
+                        tokio::fs::remove_file(&req.destination).await?;
+                        return Err(Error::Reqwest(e))
+                    },
+                }
+            }
+        }
+    }
```
The manual `drop`, `remove_file`, `req.status.send`, `req.result.send` are error prone, do we have any ideas how to automatically do these

cyberphantom52 committed:
downloader: Add better progress updates with rate limiting

cyberphantom52 committed:
downloader: Better download interface with errors

cyberphantom52 committed:
downloader: Add some convinience methods related to download status

cyberphantom52 committed:
downloader: Add config for download manager

cyberphantom52 committed:
downloader: Use TaskTracker to track thread lifecycles

cyberphantom52 committed:
downloader: Add shutdown, active_downloads and queued_downloads

cyberphantom52 committed:
downloader: Add DownloadConfig and rate limiting

cyberphantom52 committed:
downloader: Use builder pattern to build download requests

cyberphantom52 commented on src/downloader/manager.rs:
```diff
@@ -0,0 +1,139 @@
+        manager
+    }
+
+    pub fn download(
```
Should we rename this method to something like `new_request_builder` since it now returns a download request?

cyberphantom52 committed:
downloader: Don't rexport worker_thread

cyberphantom52 committed:
downloader: Switch to url::Url

This will help with swapping HTTP backends in the future

cyberphantom52 committed:
downloader: Accept Url's directly instead of parsing them

cyberphantom52 committed:
downloader: Add convinience methods to send updates and result

mirkobrombin commented on src/downloader/worker.rs:
```diff
@@ -0,0 +1,120 @@
+}
+
+async fn download(client: Client, req: &mut DownloadRequest) -> Result<File, Error> {
+    let mut response = client
+        .get(req.url().as_ref())
+        .send()
+        .await?
+        .error_for_status()?;
```
We might want to use `config.user_agent()` to set the request’s User‑Agent header, here

mirkobrombin replied:
We might want to use `config.user_agent()` to set the request’s User‑Agent header, here

mirkobrombin commented on src/downloader/request.rs:
```diff
@@ -0,0 +1,161 @@
+                status: status_tx,
+                cancel: cancel.clone(),
+                config,
+                last_progress_update: Instant::now(),
```
Might be a good idea to initialize `last_progress_update` so the first InProgress status is sent immediately?

mirkobrombin commented on src/downloader/worker.rs:
```diff
@@ -0,0 +1,120 @@
+
+    let mut last_error = None;
+    let max_attempts = req.config().max_retries();
+    for attempt in 0..=(max_attempts + 1) {
+        if attempt > max_attempts {
+            req.fail(Error::Download(DownloadError::RetriesExhausted {
+                last_error_msg: last_error
+                    .as_ref()
+                    .map(ToString::to_string)
+                    .unwrap_or_else(|| "Unknown Error".to_string()),
+            }));
+            return;
```
Maybe consider iterating `0..=max_attempts` and handling the exhausted case when `attempt == max_attempts`

cyberphantom52 replied:
soooooo while writing this what i had in mind is that  this is `max_retries` not `max_attempts` but i must've forgot to rename this.

But having attempts sounds like a better idea so i'll refactor the logic around that.

mirkobrombin commented on src/downloader/worker.rs:
```diff
@@ -0,0 +1,120 @@
+
+        if attempt > 0 {
+            req.retry();
+            // Basic exponential backoff
+            let delay_ms = 1000 * 2u64.pow(attempt as u32 - 1);
+            let delay = Duration::from_millis(delay_ms);
```
Maybe cap the exponential backoff and sprinkle in some jitter so retries don’t wait forever.

mirkobrombin commented on src/downloader/worker.rs:
```diff
@@ -0,0 +1,120 @@
+
+    let mut last_error = None;
+    let max_attempts = req.config().max_retries();
+    for attempt in 0..=(max_attempts + 1) {
+        if attempt > max_attempts {
+            req.fail(Error::Download(DownloadError::RetriesExhausted {
+                last_error_msg: last_error
+                    .as_ref()
+                    .map(ToString::to_string)
+                    .unwrap_or_else(|| "Unknown Error".to_string()),
+            }));
+            return;
+        }
+
+        if attempt > 0 {
+            req.retry();
+            // Basic exponential backoff
+            let delay_ms = 1000 * 2u64.pow(attempt as u32 - 1);
+            let delay = Duration::from_millis(delay_ms);
+
+            tokio::select! {
+                _ = tokio::time::sleep(delay) => {},
+                _ = req.cancel.cancelled() => {
+                    req.cancel();
+                    return;
+                }
+            }
+        }
+
+        match download(client.clone(), &mut req).await {
```
A quick check if the request is canceled at the start of each attempt might let us exit before starting the network request

cyberphantom52 replied:
doesn't `tokio::select` do that for us?

mirkobrombin commented on src/downloader/worker.rs:
```diff
@@ -0,0 +1,120 @@
+    loop {
+        tokio::select! {
+            _ = req.cancel.cancelled() => {
+                drop(file); // Manually drop the file handle to ensure that deletion doesn't fail
+                tokio::fs::remove_file(&req.destination()).await?;
+                return Err(Error::Download(DownloadError::Cancelled));
```
Perhaps ignore any errors from `remove_file` so cancellation doesn’t hide the real problem

Same at L106‑108

cyberphantom52 replied:
We should log the errors instead of ignoring them. We need a logger.

mirkobrombin commented on src/downloader/manager.rs:
```diff
@@ -0,0 +1,133 @@
+        self.queue.max_capacity() - self.queue.capacity()
+    }
+
+    pub fn active_downloads(&self) -> usize {
+        // -1 because the dispatcher thread is always running
+        self.tracker.len() - 1
```
Using [`saturating_sub(1)`](https://doc.rust-lang.org/std/intrinsics/fn.saturating_sub.html) could keep `active_downloads` from goinig negative if the dispatcher isn't running

cyberphantom52 replied:
Good catch

mirkobrombin commented on src/error.rs:
```diff
@@ -6,4 +7,26 @@ pub enum Error {
+    #[error("Reqwest: {0}")]
+    Reqwest(#[from] reqwest::Error),
+    #[error("Oneshot: {0}")]
+    Oneshot(#[from] tokio::sync::oneshot::error::RecvError),
+    #[error("Download: {0}")]
```
Maybe, not sure, `RecvError` could map to a more general channel cancellation variant instead of adding a dedicated Oneshot error

cyberphantom52 commented:
Superseded by https://github.com/bottlesdevs/bottles-next/pull/2
